Source/DafnyCore/Verifier/BoogieGenerator.Decreases.cs

```diff
@@ -106,28 +102,36 @@ public partial class BoogieGenerator {
   /// If builder is non-null, then the check '0 ATMOST decr' is generated to builder.
   /// Requires all types in types0 and types1 to be non-proxy non-synonym types (that is, callers should invoke NormalizeExpand)
   /// </summary>
-  Bpl.Expr DecreasesCheck(List<IToken> toks, List<Type> types0, List<Type> types1, List<Bpl.Expr> ee0, List<Bpl.Expr> ee1,
+  Bpl.Expr DecreasesCheck(List<IToken> toks, List<VarDeclStmt> prevGhostLocals,
+                          List<Expression> dafny0, List<Expression> dafny1, List<Bpl.Expr> ee0, List<Bpl.Expr> ee1,
                           BoogieStmtListBuilder builder, string suffixMsg, bool allowNoChange, bool includeLowerBound) {
     Contract.Requires(cce.NonNullElements(toks));
-    Contract.Requires(cce.NonNullElements(types0));
-    Contract.Requires(cce.NonNullElements(types1));
+    Contract.Requires(cce.NonNullElements(dafny0));
+    Contract.Requires(cce.NonNullElements(dafny1));
     Contract.Requires(cce.NonNullElements(ee0));
     Contract.Requires(cce.NonNullElements(ee1));
     Contract.Requires(predef != null);
-    Contract.Requires(types0.Count == types1.Count && types0.Count == ee0.Count && ee0.Count == ee1.Count);
+    Contract.Requires(dafny0.Count == dafny1.Count && dafny0.Count == ee0.Count && ee0.Count == ee1.Count);
     Contract.Requires(builder == null || suffixMsg != null);
     Contract.Ensures(Contract.Result<Bpl.Expr>() != null);
 
-    int N = types0.Count;
+    int N = dafny0.Count;
 
+    // TODO: NormalizeExpandKeepConstraints below?
     // compute eq and less for each component of the lexicographic tuple
     List<Bpl.Expr> Eq = new List<Bpl.Expr>(N);
     List<Bpl.Expr> Less = new List<Bpl.Expr>(N);
+    List<Expression> EqDafny = new List<Expression>(N);
+    List<Expression> LessDafny = new List<Expression>(N);
     for (int i = 0; i < N; i++) {
       Bpl.Expr less, atmost, eq;
-      ComputeLessEq(toks[i], types0[i], types1[i], ee0[i], ee1[i], out less, out atmost, out eq, includeLowerBound);
+      var ty0 = dafny0[i].Type.NormalizeExpandKeepConstraints();
+      var ty1 = dafny1[i].Type.NormalizeExpandKeepConstraints();
+      ComputeLessEq(toks[i], ty0, ty1, ee0[i], ee1[i], out less, out atmost, out eq, includeLowerBound);
       Eq.Add(eq);
+      EqDafny.Add(Expression.CreateEq(dafny0[i], dafny1[i], dafny0[i].Type.NormalizeExpand()));
       Less.Add(allowNoChange ? atmost : less);
+      LessDafny.Add(Expression.CreateLess(dafny0[i], dafny1[i]));
```

Comment:
Does this need to be conditional on `allowNoChange` like the corresponding line above?

Reply:
Yep!